cli/commands/site/delete.ts

```diff
@@ -0,0 +1,114 @@
+import fs from 'fs/promises';
+import { __, _n, sprintf } from '@wordpress/i18n';
+import { arePathsEqual } from 'common/lib/fs-utils';
+import { SiteCommandLoggerAction as LoggerAction } from 'common/logger-actions';
+import { deleteSnapshot } from 'cli/lib/api';
+import {
+	getSiteByFolder,
+	lockAppdata,
+	readAppdata,
+	saveAppdata,
+	unlockAppdata,
+	getAuthToken,
+} from 'cli/lib/appdata';
+import { connect, disconnect } from 'cli/lib/pm2-manager';
+import { stopProxyIfNoSitesNeedIt } from 'cli/lib/site-utils';
+import { getSnapshotsFromAppdata, deleteSnapshotFromAppdata } from 'cli/lib/snapshots';
+import { isServerRunning, stopWordPressServer } from 'cli/lib/wordpress-server-manager';
+import { Logger, LoggerError } from 'cli/logger';
+import { StudioArgv } from 'cli/types';
+
+export async function runCommand(
+	siteFolder: string,
+	deleteFiles: boolean = false
+): Promise< void > {
+	const logger = new Logger< LoggerAction >();
+
+	try {
+		logger.reportStart( LoggerAction.LOAD_SITES, __( 'Loading site…' ) );
+		const site = await getSiteByFolder( siteFolder, false );
+		logger.reportSuccess( __( 'Site loaded' ) );
+
+		logger.reportStart( LoggerAction.START_DAEMON, __( 'Starting process daemon...' ) );
+		await connect();
+		logger.reportSuccess( __( 'Process daemon started' ) );
+
+		const runningProcess = await isServerRunning( site.id );
+		if ( runningProcess ) {
+			logger.reportStart( LoggerAction.STOP_SITE, __( 'Stopping WordPress site...' ) );
+			await stopWordPressServer( site.id );
+			logger.reportSuccess( __( 'WordPress site stopped' ) );
+			await stopProxyIfNoSitesNeedIt( site.id, logger );
+		}
+
+		const authToken = await getAuthToken();
+		const snapshots = await getSnapshotsFromAppdata( authToken.id, siteFolder );
+
+		if ( snapshots.length > 0 ) {
+			logger.reportStart(
+				LoggerAction.DELETE_PREVIEW_SITES,
+				// translators: %d is the number of associated preview sites
+				sprintf(
+					_n(
+						'Deleting %d associated preview site…',
+						'Deleting %d associated preview sites…',
+						snapshots.length
+					),
+					snapshots.length
+				)
+			);
+
+			for ( const snapshot of snapshots ) {
+				await deleteSnapshot( snapshot.atomicSiteId, authToken.accessToken );
+				await deleteSnapshotFromAppdata( snapshot.url );
+			}
```

Comment:
```suggestion
			await Promise.all(
                  snapshots.map( ( snapshot ) =>
                  deleteSnapshot( snapshot.atomicSiteId, authToken.accessToken )
               )
           );
           
           await deleteSnapshotsFromAppdata( snapshots );
```

We could parallelize the API calls with Promise.all and create a batch deleteSnapshotsFromAppdata function to avoid multiple lock/unlock cycles on appdata.

Reply:
I've put both `deleteSnapshot` and `deleteSnapshotFromAppdata` in the `Promise.all` call. The appfile locking mechanism should prevent any issues with parallel writes, and this solution saves us from having to check which `deleteSnapshots` promises finished successfully and not before updating the appdata file.